bpf-gpl/xdp.c

```diff
@@ -65,11 +66,63 @@ int calico_xdp(struct xdp_md *xdp_ctx) {
 	}
 
 	// Parse packets and drop malformed and unsupported ones
-	if (parse_packet_ip(&ctx) == -2) {
-		return XDP_DROP;
+	switch (parse_packet_ip(&ctx)) {
+	case -1:
+		ctx.fwd.res = XDP_DROP;
+		goto deny;
+	case -2:
+		ctx.fwd.res = XDP_PASS;
+		goto allow;
 	}
 
+	tc_state_fill_from_iphdr(&ctx);
+
+	switch(tc_state_fill_from_nexthdr(&ctx)) {
+	case -1:
+		ctx.fwd.res = XDP_DROP;
+		goto deny;
+	case -2:
+		ctx.fwd.res = XDP_PASS;
+		goto allow;
+	}
+
+	// Allow a packet if it hits an entry in the failsafe map
+	if (is_failsafe_in(ctx.state->ip_proto, ctx.state->dport, ctx.state->ip_src)) {
+		CALI_DEBUG("Inbound failsafe port: %d. Skip policy\n", ctx.state->post_nat_dport);
```

Comment:
I guess it's wrong to have `dport` on the previous line and `post_nat_dport` here.  I think it should be `dport` both times, as "post-NAT" doesn't make sense here.

Reply:
right.